### PR TITLE
Add null check for target to optimized delegate constructor code path

### PR DESCRIPTION
**Pull request by joncham**
Fixes case 58888

### CODE REVIEW

**Comment by vargaz** [comment]
Looks ok. It would be better to integrate the test into another test file, these small tests have a lot of overhead, each of them takes time to compile etc.

**Comment by joncham** [comment]
I'll move test

**Comment by joncham** [comment]
Is there any strategy to where it should go? And I won't do in this PR, but any reason all `delegate*.cs` tests shouldn't just be merged?

**Comment by vargaz** [comment]
I think they should be merged and use TestDriver.cs like the other files like pinvoke2.cs.

**Comment by vargaz** [comment]
I usually put jit tests into one of the .cs files inside mini/, in this case object.cs.

File: mono/mini/method-to-ir.c [hunk 1]
```diff
@@ -3981,6 +3981,8 @@ handle_delegate_ctor (MonoCompile *cfg, MonoClass *klass, MonoInst *target, Mono
 	/* Set target field */
 	/* Optimize away setting of NULL target */
 	if (!MONO_INS_IS_PCONST_NULL (target)) {
+		MONO_EMIT_NEW_BIALU_IMM (cfg, OP_COMPARE_IMM, -1, target->dreg, 0);
```
Comment (kumpera) [inline mono/mini/method-to-ir.c]: Shouldn't this use MONO_EMIT_NULL_CHECK ?

Comment (joncham) [inline mono/mini/method-to-ir.c]: I thought so too and tried it. Turns out it doesn't do anything unless explicit null checks are enabled:
https://github.com/mono/mono/blob/492438a431ea5bf563735055441cbe81a453b195/mono/mini/ir-emit.h#L889

**Pull request merged**